thorin/fe/lexer.cpp

```diff
@@ -39,16 +39,15 @@ Tok Lexer::lex() {
             return *cache;
         }
 
-        loc_.begin = ahead().pos;
-        str_.clear();
+        begin();
 
 #if defined(_WIN32) && !defined(NDEBUG) // isspace asserts otherwise
         if (accept_if([](int c) { return (c & ~0xFF) == 0 ? isspace(c) : false; })) continue;
 #else
         if (accept_if(isspace)) continue;
 #endif
-        if (accept(utf8::Err)) error(loc_, "invalid UTF-8 character");
-        if (accept(utf8::EoF)) return tok(Tag::M_eof);
+        if (accept(0)) error(loc_, "invalid UTF-8 character");
```

Comment:
why introduce a new magic constant when you had an enum for it?
if you mean `\0`, maybe use that?